File: tests/test_appeal_coaching.py
from pathlib import Path
import sys
import unittest

sys.path.insert(0, str(Path(__file__).resolve().parents[1] / "src"))

from appeal.coaching import (
    appeal_coaching_hint,
    appeal_coaching_verdicts,
    appeal_evidence_outcome,
    appeal_judge_outcomes,
    appeal_score_movement,
    assess_appeal_coaching,
    is_degenerate_evidence_asks,
    normalize_target_judges,
)
from judges.schemas import RoastPanel, Verdict, VerdictLabel, judgeLabel
from verification.invariants import is_generic_evidence

SAMPLE_EVIDENCE = "Three signed LOIs from target buyers would change this verdict."


def _verdict(
    judge: judgeLabel,
    *,
    verdict: VerdictLabel,
    score: int,
    key_concern: str,
    evidence: str | None = None,
) -> Verdict:
    return Verdict(
        judge=judge,
        verdict=verdict,
        roast="This pitch still leaves material gaps the founder must close.",
        score=score,
        key_concern=key_concern,
        evidence_to_change_verdict=evidence,
    )


class AppealCoachingTests(unittest.TestCase):
    def test_uses_evidence_when_present(self):
        verdict = _verdict(
            judgeLabel.VC,
            verdict=VerdictLabel.CONDITIONAL,
            score=5,
            key_concern="No repeatable GTM motion yet.",
            evidence=SAMPLE_EVIDENCE,
        )
        self.assertEqual(appeal_coaching_hint(verdict), SAMPLE_EVIDENCE)

    def test_falls_back_to_key_concern(self):
        verdict = _verdict(
            judgeLabel.CUSTOMER,
            verdict=VerdictLabel.FAIL,
            score=2,
            key_concern="No proof buyers will switch from incumbents.",
        )
        hint = appeal_coaching_hint(verdict)
        self.assertIn("No proof buyers will switch from incumbents.", hint)
        self.assertNotIn(SAMPLE_EVIDENCE, hint)

    def test_whitespace_only_evidence_falls_back_to_key_concern(self):
        verdict = Verdict.model_construct(
            judge=judgeLabel.VC,
            verdict=VerdictLabel.FAIL,
            roast="This pitch still leaves material gaps the founder must close.",
            score=2,
            key_concern="No traction signal yet.",
            evidence_to_change_verdict="   ",
        )
        self.assertIn("No traction signal yet.", appeal_coaching_hint(verdict))

    def test_sorts_fail_and_conditional_before_pass(self):
        panel = RoastPanel(
            verdicts=[
                _verdict(
                    judgeLabel.VC,
                    verdict=VerdictLabel.PASS,
                    score=8,
                    key_concern="Minor scale risk.",
                    evidence="Already a pass.",
                ),
                _verdict(
                    judgeLabel.ENGINEER,
                    verdict=VerdictLabel.FAIL,
                    score=2,
                    key_concern="No technical moat.",
                    evidence="Pilot with 95% accuracy.",
                ),
                _verdict(
                    judgeLabel.PM,
                    verdict=VerdictLabel.CONDITIONAL,
                    score=5,
                    key_concern="Unclear wedge.",
                    evidence="One repeatable channel with conversion data.",
                ),
                _verdict(
                    judgeLabel.CUSTOMER,
                    verdict=VerdictLabel.PASS,
                    score=7,
                    key_concern="Onboarding friction remains.",
                    evidence="Usability test with three successful completions.",
                ),
                _verdict(
                    judgeLabel.COMPETITOR,
                    verdict=VerdictLabel.CONDITIONAL,
                    score=4,
                    key_concern="Incumbents can copy quickly.",
                    evidence="Exclusive data partnership blocking replication.",
                ),
            ]
        )
        ordered = appeal_coaching_verdicts(panel)
        self.assertEqual(
            [verdict.judge for verdict in ordered],
            [
                judgeLabel.ENGINEER,
                judgeLabel.COMPETITOR,
                judgeLabel.PM,
                judgeLabel.CUSTOMER,
                judgeLabel.VC,
            ],
        )
        self.assertEqual(
            [verdict.verdict for verdict in ordered],
            [
                VerdictLabel.FAIL,
                VerdictLabel.CONDITIONAL,
                VerdictLabel.CONDITIONAL,
                VerdictLabel.PASS,
                VerdictLabel.PASS,
            ],
        )


class AppealCoachingPhase2Tests(unittest.TestCase):
    def test_normalize_target_judges_keeps_panel_order(self):
        self.assertEqual(
            normalize_target_judges(["customer", "vc", "unknown"]),
            ("vc", "customer"),
        )

    def test_evidence_outcome_uses_score_delta(self):
        original = _verdict(
            judgeLabel.VC,
            verdict=VerdictLabel.CONDITIONAL,
            score=4,
            key_concern="No LOIs yet.",
            evidence=SAMPLE_EVIDENCE,
        )
        raised = original.model_copy(update={"score": 6})
        unchanged = original.model_copy(update={"score": 4})
        self.assertEqual(appeal_evidence_outcome(original, raised), "Evidence met")
        self.assertEqual(appeal_evidence_outcome(original, unchanged), "Not met")

    def test_evidence_outcome_for_pass_and_negative_delta(self):
        passing = _verdict(
            judgeLabel.VC,
            verdict=VerdictLabel.PASS,
            score=8,
            key_concern="Minor scale risk.",
        )
        self.assertEqual(
            appeal_evidence_outcome(passing, passing.model_copy(update={"score": 8})),
            "Already passing",
        )
        lowered = passing.model_copy(update={"score": 7})
        self.assertEqual(appeal_evidence_outcome(passing, lowered), "Not met")

    def test_appeal_judge_outcomes_include_targeting(self):
        baseline = RoastPanel(
            verdicts=[
                _verdict(
                    judgeLabel.VC,
                    verdict=VerdictLabel.CONDITIONAL,
                    score=4,
                    key_concern="No LOIs yet.",
                    evidence=SAMPLE_EVIDENCE,
                ),
                _verdict(
                    judgeLabel.ENGINEER,
                    verdict=VerdictLabel.FAIL,
                    score=2,
                    key_concern="No technical moat.",
                ),
                _verdict(
                    judgeLabel.PM,
                    verdict=VerdictLabel.CONDITIONAL,
                    score=5,
                    key_concern="Unclear wedge.",
                ),
                _verdict(
                    judgeLabel.CUSTOMER,
                    verdict=VerdictLabel.FAIL,
                    score=3,
                    key_concern="No buyer proof.",
                ),
                _verdict(
                    judgeLabel.COMPETITOR,
                    verdict=VerdictLabel.CONDITIONAL,
                    score=4,
                    key_concern="Easy to copy.",
                ),
            ]
        )
        revised = baseline.model_copy(
            update={
                "verdicts": [
                    baseline.verdicts[0].model_copy(update={"score": 6}),
                    *baseline.verdicts[1:],
                ]
            }
        )
        outcomes = appeal_judge_outcomes(baseline, revised, ("vc",))
        vc_outcome = next(item for item in outcomes if item.judge == "vc")
        self.assertTrue(vc_outcome.targeted)
        self.assertEqual(vc_outcome.outcome, "Evidence met")
        self.assertEqual(vc_outcome.evidence_ask, SAMPLE_EVIDENCE)


class AppealCoachingPhase3Tests(unittest.TestCase):
    def test_assess_flags_degenerate_asks(self):
        shared = "Provide more evidence and do more research."
        panel = RoastPanel(
            verdicts=[
                _verdict(
                    judgeLabel.VC,
                    verdict=VerdictLabel.FAIL,
                    score=2,
                    key_concern="No traction.",
                    evidence=shared,
                ),
                _verdict(
                    judgeLabel.ENGINEER,
                    verdict=VerdictLabel.FAIL,
                    score=2,
                    key_concern="No moat.",
                    evidence=shared,
                ),
                _verdict(
                    judgeLabel.PM,
                    verdict=VerdictLabel.CONDITIONAL,
                    score=5,
                    key_concern="Unclear wedge.",
                    evidence=shared,
                ),
                _verdict(
                    judgeLabel.CUSTOMER,
                    verdict=VerdictLabel.CONDITIONAL,
                    score=4,
                    key_concern="No buyer proof.",
                    evidence=shared,
                ),
                _verdict(
                    judgeLabel.COMPETITOR,
                    verdict=VerdictLabel.CONDITIONAL,
                    score=4,
                    key_concern="Easy to copy.",
                    evidence=shared,
                ),
            ]
        )
        coaching = assess_appeal_coaching(panel)
        self.assertTrue(coaching["degraded"])
        self.assertTrue(coaching["degenerate_asks"])
        self.assertTrue(is_degenerate_evidence_asks(panel.verdicts))

    def test_assess_flags_generic_and_derived_asks(self):
        panel = RoastPanel(
            verdicts=[
                _verdict(
                    judgeLabel.VC,
                    verdict=VerdictLabel.FAIL,
                    score=2,
                    key_concern="No LOIs yet.",
                    evidence="Show traction with signed LOIs.",
                ),
                _verdict(
                    judgeLabel.ENGINEER,
                    verdict=VerdictLabel.FAIL,
                    score=2,
                    key_concern="No technical moat.",
                ),
                _verdict(
                    judgeLabel.PM,
                    verdict=VerdictLabel.CONDITIONAL,
                    score=5,
                    key_concern="Unclear wedge.",
                    evidence="Do more research on the buyer.",
                ),
                _verdict(
                    judgeLabel.CUSTOMER,
                    verdict=VerdictLabel.CONDITIONAL,
                    score=4,
                    key_concern="No buyer proof.",
                    evidence=SAMPLE_EVIDENCE,
                ),
                _verdict(
                    judgeLabel.COMPETITOR,
                    verdict=VerdictLabel.CONDITIONAL,
                    score=4,
                    key_concern="Easy to copy.",
                    evidence=SAMPLE_EVIDENCE,
                ),
            ]
        )
        coaching = assess_appeal_coaching(panel)
        qualities = {item.judge: item.quality for item in coaching["items"]}
        self.assertEqual(qualities["vc"], "precise")
        self.assertEqual(qualities["engineer"], "derived")
        self.assertEqual(qualities["pm"], "generic")
        self.assertEqual(qualities["customer"], "duplicate")
        self.assertTrue(coaching["degraded"])
        self.assertFalse(is_generic_evidence("Show traction with signed LOIs."))
        self.assertTrue(is_generic_evidence("Do more research on the buyer."))

    def test_single_derived_ask_does_not_degrade_banner(self):
        panel = RoastPanel(
            verdicts=[
                _verdict(
                    judgeLabel.VC,
                    verdict=VerdictLabel.FAIL,
                    score=2,
                    key_concern="No LOIs yet.",
                    evidence=SAMPLE_EVIDENCE,
                ),
                _verdict(
                    judgeLabel.ENGINEER,
                    verdict=VerdictLabel.FAIL,
                    score=2,
                    key_concern="No technical moat.",
                ),
                _verdict(
                    judgeLabel.PM,
                    verdict=VerdictLabel.CONDITIONAL,
                    score=5,
                    key_concern="Unclear wedge.",
                    evidence="One repeatable channel with conversion data.",
                ),
                _verdict(
                    judgeLabel.CUSTOMER,
                    verdict=VerdictLabel.CONDITIONAL,
                    score=4,
                    key_concern="No buyer proof.",
                    evidence="Usability test with three successful completions.",
                ),
                _verdict(
                    judgeLabel.COMPETITOR,
                    verdict=VerdictLabel.CONDITIONAL,
                    score=4,
                    key_concern="Easy to copy.",
                    evidence="Exclusive data partnership blocking replication.",
                ),
            ]
        )
        coaching = assess_appeal_coaching(panel)
        self.assertFalse(coaching["degraded"])

    def test_appeal_score_movement_counts_positive_moves(self):
        baseline = RoastPanel(
            verdicts=[
                _verdict(
                    judgeLabel.VC,
                    verdict=VerdictLabel.CONDITIONAL,
                    score=4,
                    key_concern="No LOIs yet.",
                    evidence=SAMPLE_EVIDENCE,
                ),
                _verdict(
                    judgeLabel.ENGINEER,
                    verdict=VerdictLabel.FAIL,
                    score=2,
                    key_concern="No technical moat.",
                ),
                _verdict(
                    judgeLabel.PM,
                    verdict=VerdictLabel.CONDITIONAL,
                    score=5,
                    key_concern="Unclear wedge.",
                ),
                _verdict(
                    judgeLabel.CUSTOMER,
                    verdict=VerdictLabel.FAIL,
                    score=3,
                    key_concern="No buyer proof.",
                ),
                _verdict(
                    judgeLabel.COMPETITOR,
                    verdict=VerdictLabel.CONDITIONAL,
                    score=4,
                    key_concern="Easy to copy.",
                ),
            ]
        )
        revised = baseline.model_copy(
            update={
                "verdicts": [
                    baseline.verdicts[0].model_copy(update={"score": 6}),
                    baseline.verdicts[1],
                    baseline.verdicts[2].model_copy(update={"score": 6}),
                    baseline.verdicts[3],
                    baseline.verdicts[4],
                ]
            }
        )
        movement = appeal_score_movement(baseline, revised)
        self.assertEqual(movement["positive_moves"], 2)
        self.assertEqual(movement["net_delta"], 3)


if __name__ == "__main__":
    unittest.main()

File: web/scripts/test-appeal-coaching.mjs
import assert from "node:assert/strict";
import test from "node:test";

import {
  appealEvidenceOutcome,
  appealJudgeOutcomes,
  appealScoreMovement,
  assessAppealCoaching,
  deriveTargetJudgesForEvidence,
  findDuplicateEvidenceJudges,
  isGenericEvidence,
  normalizeTargetJudges,
  panelAverageScore,
  summarizeEvidenceOutcomes,
} from "../src/lib/appeal/coaching.ts";

const BASELINE = {
  judge: "vc",
  verdict: "CONDITIONAL",
  roast: "Needs proof.",
  score: 4,
  key_concern: "No LOIs yet.",
  evidence_to_change_verdict: "Three signed LOIs would change this verdict.",
};

test("normalizeTargetJudges keeps panel order and drops unknown ids", () => {
  assert.deepEqual(normalizeTargetJudges(["customer", "vc", "unknown"]), ["vc", "customer"]);
  assert.deepEqual(normalizeTargetJudges(undefined), []);
});

test("appealEvidenceOutcome follows score delta", () => {
  assert.equal(
    appealEvidenceOutcome(BASELINE, { ...BASELINE, score: 6 }),
    "Evidence met",
  );
  assert.equal(
    appealEvidenceOutcome(BASELINE, { ...BASELINE, score: 4 }),
    "Not met",
  );
  assert.equal(
    appealEvidenceOutcome(
      { ...BASELINE, verdict: "PASS", score: 8 },
      { ...BASELINE, verdict: "PASS", score: 8 },
    ),
    "Already passing",
  );
});

test("appealJudgeOutcomes marks targeted judges", () => {
  const revised = { ...BASELINE, score: 6 };
  const outcomes = appealJudgeOutcomes([BASELINE], [revised], ["vc"]);
  assert.equal(outcomes.length, 1);
  assert.equal(outcomes[0]?.targeted, true);
  assert.equal(outcomes[0]?.outcome, "Evidence met");
});

test("assessAppealCoaching flags generic and duplicate asks", () => {
  const verdicts = [
    { ...BASELINE, judge: "vc", evidence_to_change_verdict: "Show traction with signed LOIs." },
    { ...BASELINE, judge: "engineer", key_concern: "No moat.", evidence_to_change_verdict: undefined },
    { ...BASELINE, judge: "pm", evidence_to_change_verdict: "Do more research on the buyer." },
    { ...BASELINE, judge: "customer", evidence_to_change_verdict: "Three signed LOIs would change this verdict." },
    { ...BASELINE, judge: "competitor", evidence_to_change_verdict: "Three signed LOIs would change this verdict." },
  ];
  const coaching = assessAppealCoaching(verdicts);
  assert.equal(coaching.degraded, true);
  assert.equal(coaching.items.find((item) => item.judge === "engineer")?.quality, "derived");
  assert.equal(coaching.items.find((item) => item.judge === "pm")?.quality, "generic");
  assert.equal(coaching.items.find((item) => item.judge === "customer")?.quality, "duplicate");
  assert.equal(findDuplicateEvidenceJudges(verdicts).has("competitor"), true);
  assert.equal(isGenericEvidence("Show traction with signed LOIs."), false);
  assert.equal(isGenericEvidence("Do more research on the buyer."), true);
});

test("appealScoreMovement counts positive moves", () => {
  const baseline = [
    { ...BASELINE, judge: "vc", score: 4 },
    { ...BASELINE, judge: "engineer", score: 4 },
  ];
  const revised = [
    { ...BASELINE, judge: "vc", score: 6 },
    { ...BASELINE, judge: "engineer", score: 4 },
  ];
  assert.deepEqual(appealScoreMovement(baseline, revised), {
    positiveMoves: 1,
    netDelta: 2,
  });
});

test("deriveTargetJudgesForEvidence maps recommended fix to judge", () => {
  const verdicts = [
    { ...BASELINE, judge: "vc", recommended_fix: "Interview ten buyers." },
    { ...BASELINE, judge: "pm", recommended_fix: "Narrow the wedge." },
  ];
  assert.deepEqual(deriveTargetJudgesForEvidence(verdicts, "Interview ten buyers."), ["vc"]);
});

test("deriveTargetJudgesForEvidence falls back to worst judge", () => {
  const verdicts = [
    { ...BASELINE, judge: "vc", verdict: "PASS", score: 8 },
    { ...BASELINE, judge: "pm", verdict: "FAIL", score: 2 },
  ];
  assert.deepEqual(deriveTargetJudgesForEvidence(verdicts, "Unknown risk."), ["pm"]);
});

test("deriveTargetJudgesForEvidence ignores synthesis-only problem text", () => {
  const verdicts = [
    { ...BASELINE, judge: "vc", verdict: "FAIL", score: 2, recommended_fix: "Interview ten buyers." },
    { ...BASELINE, judge: "pm", verdict: "CONDITIONAL", score: 5, recommended_fix: "Narrow the wedge." },
  ];
  assert.deepEqual(
    deriveTargetJudgesForEvidence(verdicts, "No buyer proof yet."),
    ["vc"],
  );
});

test("panelAverageScore averages judge scores", () => {
  assert.equal(
    panelAverageScore([
      { ...BASELINE, score: 4 },
      { ...BASELINE, judge: "pm", score: 6 },
    ]),
    5,
  );
});

test("summarizeEvidenceOutcomes describes targeted outcomes", () => {
  const outcomes = appealJudgeOutcomes(
    [BASELINE],
    [{ ...BASELINE, score: 6 }],
    ["vc"],
  );
  assert.match(summarizeEvidenceOutcomes(outcomes), /targeted judge.*met/i);
});
